docs: Add content to KFP docs and remove MinIO sharing warning (#177)
StatCan/daaas: docs/en/3-Pipelines/Kubeflow-Pipelines.md

```

In this YAML we see the input parameters passed are placeholders for data from
previous components rather than their actual value. This is because while KFP
knows **a** result from `average-data` and `average-2-data` will be passed to
average, but that the **value** of that result is not available until the
pipeline is actually run.

<!-- prettier-ignore -->
??? info "Component naming within the YAML file"
    Because we made a `average_op` factory function with `name='average'` above,
     our YAML file has component names that automatically increment to avoid 
     recreating the same name twice.  We could have been fancier with our 
     factory functions to more directly control our names, giving an argument 
     like `name='average_first_input_args'`, or could even have explicitly 
     defined the name in our pipeline by using 
     `avg_1 = average_op(a, b, c).set_display_name("Average 1")`

As one more example, let's try two more pipelines. One has a for loop inside
which prints "Woohoo!" a fixed number of times. whereas the other does the same
but loops `n` times (where n is a pipeline parameter):

<!-- prettier-ignore -->

> !!! info "Pipeline parameters are described more below, but they work like 

Author: sorry not sure what you mean by indentation.  I didn't see anything when looking at the doc

KFP cannot define a pipeline from this because it does not know how many times
to loop. We'd hit similar problems if using `if` statements. There are some ways
around this, but they're left to the reader to explore through the
[Kubeflow Pipelines docs](https://www.kubeflow.org/docs/pipelines/overview/pipelines-overview/).

Why does pipeline authoring behave this way? Because _pipelines_ (and
_components_)) are meant to be reusable definitions of logic that are defined in
static YAML files, with all dynamic decision making done inside components. This
can make them a little awkward to define, but also helps them be more reusable.

# Data Exchange

## Passing data into, within, and from a pipeline

In the first example above, we pass:

- numbers into our pipeline
- numbers between components within our pipeline
- a number back to the user at the end

But as discussed above, pipeline arguments and component results are just
placeholder objects - so how does KFP know our values are numeric? The answer
is: it doesn't. In fact, it didn't even treat them as numbers above. Instead it
treated them as strings. It is just that our pipeline components worked just as

> well with `'5'` as they would have with `5`.

Author: My first reaction was "this seems picky".  But your explanation makes total sense hahaha

which are components that can be built straight from python without building new
container images for each change. These components are great for fast iteration
during development, as well as for simple tasks that can be written and managed
easily.

This is an example of a lightweight pipeline with a single component that
concatenates strings:

```python
import kfp
from kfp import dsl
from kfp.components import func_to_container_op

def concat_string(a, b) -> str:
    return f"({a} | {b})"

concat_string_component = func_to_container_op(concat_string,
                                               base_image="python:3.8.3-buster"
                                               )
@dsl.pipeline(
    name="My lightweight pipeline",
)
def pipeline(str1, str2, str3):
    # Note that we use the concat_string_component, not the
    # original concat_string() function
    concat_result_1 = concat_string_component(str1, str2)

    # By using cancat_result_1's output, we define the dependency of
    # concat_result_2 on concat_result_1
    concat_result_2 = concat_string_component(concat_result_1.output, str3)
```

> where we see our `concat_string` component is defined directly in python rather
> than from a Docker image. In the end our function still runs in a container, but
> instead of us building a new image the `func_to_container_op()` runs our python
> code inside the provided base image (`python:3.8.3-buster`). This lets us avoid
> authoring a new image each iteration. The base image can be anything accessible

Reviewer: ```suggestion
We see that our `concat_string` component is defined directly in Python rather
than from a Docker image. In the end, our function still runs in a container, but
we don't have to build it ourselves: `func_to_container_op()` runs our Python
code inside the provided base image (`python:3.8.3-buster`). This lets us avoid
building a new image every time we change our code. The base image can be anything accessible
```

by Kubeflow, which includes all images in the Azure Container Registry and any
[whitelisted images](https://github.com/StatCan/gatekeeper-policies/blob/master/general/container-allowed-images/constraint.yaml)
from Docker hub.

<!-- prettier-ignore -->
??? danger "Lightweight components have a number of advantages but also some drawbacks"

>     See [this description](https://github.com/StatCan/jupyter-notebooks/blob/master/kfp-basics/demo_kfp_lightweight_components.ipynb) of their basic characteristics, as well as [this example](https://github.com/StatCan/jupyter-notebooks/blob/master/mapreduce-pipeline/Compute-Pi-with-lightweight-components-and-minio.ipynb) which uses them in a more complex pipeline

Reviewer: Line length

<!-- prettier-ignore -->
??? info "A convenient base image to use is the the image your notebook server is running"

>     By using the same image as your notebook server, you ensure Kubeflow Pipelines has the same packages available to it as the notebook you author from.  This can help avoid accidentally importing packages you have locally but not in Kubeflow.  You can find that link from the notebook server page as shown below, but make sure you prepend the registry URL (so the below image would have `base_image=k8scc01covidacr.azurecr.io/machine-learning-notebook-cpu:562fa4a2899eeb9ae345c51c2491447ec31a87d7`).  Note though that while using a fully featured base image for iteration is fine, for production pipelines it is good to keep these images lean and only supply the necessary software.  That way you reduce the startup time for each step in your pipeline.

Reviewer: ```suggestion
    By using the same image as your notebook server, you ensure Kubeflow Pipelines has the same packages available to it as the notebook where you do your analysis. This can help avoid errors from importing packages specific specific to your environment. You can find that link from the notebook server page as shown below, but make sure you prepend the registry URL (so the below image would have `base_image=k8scc01covidacr.azurecr.io/machine-learning-notebook-cpu:562fa4a2899eeb9ae345c51c2491447ec31a87d7`). Note that while using a fully featured base image for iteration is fine, it's a good practice to keep production pipelines lean and only supply the necessary software. That way you reduce the startup time for each step in your pipeline.
```

And line length.

    ![A Kubeflow Pipeline](../images/kubeflow_pipelines_notebook_server_image.png)

### Defining components directly in YAML

Components can be defined directly with a YAML file, where the designer can run
terminal commands from a given Docker image. This can be a great way to make

> components that use non-python code from existing containers. Like with all

Reviewer: ```suggestion
non-Python pipeline components from existing containers. As with all
```

components, we can pass both arguments and data files into/out of the component.
For example:

```yaml
name: Concat Strings
inputs:
  - {
      name: Input text 1,
      type: String,
      description: "Some text to echo into a terminal and tee to a file",
    }
  - {
      name: Input text 2,
      type: String,
      description: "Some text to echo into a terminal and tee to a file",
    }
outputs:
  - { name: Output filename, type: String }
implementation:
  container:
    image: bash:5
    command:
      - bash
      - -ex
      - -c
      - |
        echo "$0 | $1" | tee $2
      - { inputValue: Input text 1 }
      - { inputValue: Input text 2 }
      - { outputPath: Output filename }
```

> concatenates two stings similar to our lightweight example above. We then define
> a component in python from this YAML by:

Reviewer: ```suggestion
This example concatenates two strings like our lightweight example above. We then define
a component in python from this YAML:
```

```python
from kfp.components import load_component_from_file
echo_and_tee = load_component_from_file('path/to/echo_and_tee.yaml')

@dsl.pipeline
def my_pipeline():
    echo_and_tee_task_1 = echo_and_tee("My text to echo")

    # A second use that consumes the return of the first one
    echo_and_tee_task_2 = echo_and_tee(echo_and_tee_task_1.output)
```

See
[this example](https://github.com/StatCan/jupyter-notebooks/blob/master/mapreduce-pipeline/Compute-Pi-with-reusable-components-and-minio.ipynb)
for more details on using existing components.

### Reusing existing components

> Similar to well abstracted functions, well abstracted components can reduce the
> amount of code you have to do on any given project. For example, rather than
> teaching your machine learning `train_model` component to then save the model to
> MinIO, you can instead have `train_model` return the model and then Kubeflow
> Pipelines pass the model to a reusable `copy_to_minio` component. This reuse
> pattern applies to components defined through any means (containers,
> lightweight, or YAML). An example of this is given
> [here](https://github.com/StatCan/jupyter-notebooks/blob/master/mapreduce-pipeline/Compute-Pi-with-reusable-components-and-minio.ipynb),
> which reuses provided components for simple file IO tasks.

Reviewer: ```suggestion
Similar to well abstracted functions, well abstracted components can reduce the
amount of code you have to write for any given project. For example, rather than
teaching your machine learning `train_model` component to also save the resulting model to
MinIO, you can instead have `train_model` return the model and then Kubeflow
Pipelines can pass the model to a reusable `copy_to_minio` component. This reuse
pattern applies to components defined through any means (containers,
lightweight, or YAML). [Take a look at our example notebook](https://github.com/StatCan/jupyter-notebooks/blob/master/mapreduce-pipeline/Compute-Pi-with-reusable-components-and-minio.ipynb),
which reuses provided components for simple file IO tasks.
```
